provider/oci/instance.go

```diff
@@ -4,44 +4,282 @@
 package oci
 
 import (
+	"context"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
 	"github.com/juju/errors"
+	"github.com/juju/juju/status"
 
-	"github.com/juju/juju/environs/context"
+	envcontext "github.com/juju/juju/environs/context"
+	"github.com/juju/juju/environs/instances"
 	"github.com/juju/juju/instance"
 	"github.com/juju/juju/network"
+
+	ociCore "github.com/oracle/oci-go-sdk/core"
+)
+
+const (
+	// MinVolumeSizeMB is the minimum size in MB for a volume or boot disk
+	MinVolumeSizeMB = 51200
+
+	// MaxVolumeSizeMB is the maximum size in MB for a volume or boot disk
+	MaxVolumeSizeMB = 16777216
 )
 
-type ociInstance struct{}
+type ociInstance struct {
+	arch     string
+	instType *instances.InstanceType
+	env      *Environ
+	mutex    sync.Mutex
+	etag     *string
+	raw      ociCore.Instance
+}
+
+type vnicWithIndex struct {
+	Vnic ociCore.Vnic
+	Idx  int
+}
 
 var _ instance.Instance = (*ociInstance)(nil)
-var _ instance.InstanceFirewaller = (*ociInstance)(nil)
+
+// newInstance returns a new oracleInstance
+func newInstance(raw ociCore.Instance, env *Environ) (*ociInstance, error) {
+	if raw.Id == nil {
+		return nil, errors.New(
+			"Instance response does not contain an ID",
+		)
+	}
+	instance := &ociInstance{
+		raw:  raw,
+		env:  env,
+		arch: "amd64",
+	}
+
+	return instance, nil
+}
+
+// SetInstance sets the raw property of ociInstance{}
+// Testing purposes.
+func (o *ociInstance) SetInstance(inst ociCore.Instance) {
+	o.raw = inst
+}
+
+func (o *ociInstance) availabilityZone() string {
+	return *o.raw.AvailabilityDomain
+}
 
 // Id implements instance.Instance
 func (o *ociInstance) Id() instance.Id {
-	return instance.Id(0)
+	return instance.Id(*o.raw.Id)
 }
 
 // Status implements instance.Instance
-func (o *ociInstance) Status(ctx context.ProviderCallContext) instance.InstanceStatus {
-	return instance.InstanceStatus{}
+func (o *ociInstance) Status(ctx envcontext.ProviderCallContext) instance.InstanceStatus {
+	// This should not happen, unless someone bypassed newInstance()
+	// and created the ociInstance{} object manually.
+	if err := o.refresh(); err != nil {
+		return instance.InstanceStatus{}
+	}
+	return instance.InstanceStatus{
+		Status:  status.Status(o.raw.LifecycleState),
```

Comment:
I think mapping PROVISIONING, CREATING_IMAGE, and STARTING to "Pending" (allocating?) and then the rest to Running but including the Message as the more interesting state is best.
That allows people that have scripts that check instance state to continue using the Status field as the machine readable field but still gives the extra detail that the provider offers.